File: Week_04-BFS-DFS-二分-贪心/18.四数之和.cpp
/* 5.25
 * @lc app=leetcode.cn id=18 lang=cpp
 *
 * [18] 四数之和
 */
/**
 * 三数之和的升级版，多一个for循环，注意处理重复数字
 * 处理重复时，避免[0,0,0,0]被过滤掉
 * 
 **/

// @lc code=start
#include<vector>
using namespace std;
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        int size_ = nums.size();
        for (int a = 0; a < size_-3; a++) {
            if (a>0 && nums[a]==nums[a-1]) continue; //处理重复数字
            for (int b=a+1; b<size_-2; b++) {
                if (b>a+1 && nums[b]==nums[b-1]) continue;//处理重复数字
                int c = b+1, d = size_-1;
                while(c < d){
                    int sum = nums[a] + nums[b] + nums[c] + nums[d];
                    if (sum > target) d--;
                    else if (sum < target) c++;
                    else {
                        ans.push_back({nums[a], nums[b], nums[c], nums[d]});
                        while (c<d && nums[c]==nums[++c]) continue;//处理重复数字
                        while (c<d && nums[d]==nums[--d]) continue;//处理重复数字
                    }
                }
            }
        }
        return ans;
    }
};
// @lc code=end

File: Week_04-BFS-DFS-二分-贪心/18.四数之和.py
#
# @lc app=leetcode.cn id=18 lang=python
#
# [18] 四数之和
#

# @lc code=start
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        ans = list()
        nums.sort()
        for a in range(len(nums)):
            if a > 0 and nums[a] == nums[a-1]: continue
            for b in range(a+1, len(nums)-2):
                if b > a+1 and nums[b] == nums[b-1]: continue
                c = b+1
                d = len(nums) - 1
                while c < d:
                    sum = nums[a] + nums[b] + nums[c] + nums[d]
                    if sum > target:
                        d -= 1
                    if sum < target:
                        c += 1
                    #注意如何组织代码来处理重复
                    if sum == target:
                        ans.append([nums[a], nums[b], nums[c], nums[d]])
                        c += 1
                        while c < d and nums[c] == nums[c-1]: c += 1
                        d -= 1
                        while c < d and nums[d] == nums[d+1]: d -= 1
        return ans
                        

# @lc code=end

